leaves/table.ml

```diff
@@ -0,0 +1,104 @@
+type column = { title : string; width : int }
+type row = string list
+
+type styles = {
+  cursor : Spices.style;
+  header : Spices.style;
+  cell : Spices.style;
+}
+
+let default_styles =
+  {
+    cursor =
+      Spices.(
+        default |> bold true |> fg (color "#FFFFFF") |> bg (color "#053d37"));
+    header =
+      Spices.(
+        default |> bold true
+        |> fg (color "#FFFFFF")
+        |> padding_right 1 |> padding_left 1);
+    cell = Spices.(default |> padding_right 1 |> padding_left 1);
+  }
+
+type t = {
+  columns : column array;
+  rows : row list;
+  styles : styles;
+  cursor : int;
+  start_of_frame : int;
+  height_of_frame : int;
+}
+
+let update t (e : Minttea.Event.t) =
+  let go_up t steps =
+    let cursor = max (t.cursor - steps) 0 in
+    let start_of_frame =
+      if cursor < t.start_of_frame then cursor else t.start_of_frame
+    in
+    { t with cursor; start_of_frame }
+  in
+  let go_down t steps =
+    let cursor = min (t.cursor + steps) (List.length t.rows - 1) in
+    let start_of_frame =
+      if cursor > t.start_of_frame + t.height_of_frame - 1 then
+        cursor - t.height_of_frame + 1
+      else t.start_of_frame
+    in
+    { t with cursor; start_of_frame }
+  in
+  match e with
+  | Minttea.Event.KeyDown (Key "b") -> go_up t t.height_of_frame
+  | KeyDown (Key "f") | KeyDown Space -> go_down t t.height_of_frame
+  | KeyDown Up -> go_up t 1
+  | KeyDown Down -> go_down t 1
+  | _ -> t
+
+let pad_right_or_ellipsis width str =
+  (* TODO(@sabine): we probably don't want the uuseg dependency
+     or move this somewhere where it actually belongs. Also for uuseg_string_sub,
+     we need to write actually good code. *)
+  let uuseg_string_length s =
+    Uuseg_string.fold_utf_8 `Grapheme_cluster (fun len _ -> len + 1) 0 s
+  in
```

Comment:
Just a comment from a side observer but I'll be perfectly fine with having `uuseg` in dependencies 🙂 
In fact, it might already be there due to some transitive dependencies (it's definitely in my switch because of dune and utop).

I've written exactly the same function because I want to calculate sizes of TUI components precisely, so I wouldn't even mind having this function exposed via some API from Mint Tea (probably not from a table widget)

Reply:
It wasn't in my switch, that's why I was wondering if I can/should use it. But it was the quickest way to get the padding and truncating to work. :)

Reply:
I guess this could be added to `spices.ml` as a style. :thinking:

There already is a `width` style there that isn't currently implemented and could be expanded by allowing an optional `ellipsis`. However, we'd need to consider the interaction with margin and padding since that will likely not be obvious.